### DISCUSSION TITLE
Repo Setup Issues

### DISCUSSION THREAD

**Discussion by lpm0073** [FullStackWithLawrence/aws-openai#19] [General]

**Comment by jlombard** [comment]
Hello, it seems you've put a lot of effort into making this project legit, thank you!

I'm looking for help. It's probably a user error since all of this is pretty new to me, but it seems like I'm pretty close to the finish line but am now getting the following error when running `terraform apply`.   It seems that these integrations are not defined.

—- 

Plan: 17 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + api_apigateway_url           = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_api_gateway_deployment.openai: Creating...
aws_lambda_layer_version.genai: Creating...
╷
│ Error: creating API Gateway Deployment: BadRequestException: No integration defined for method
│ 
│   with aws_api_gateway_deployment.openai,
│   on apigateway.tf line 61, in resource "aws_api_gateway_deployment" "openai":
│   61: resource "aws_api_gateway_deployment" "openai" {
│ 
╵
╷
│ Error: creating lambda layer: RequestEntityTooLargeException: Request must be smaller than 70167211 bytes for the PublishLayerVersion operation
│ 	status code: 413, request id: 688a39f9-dddd-4541-8453-25ec99b27a10
│ 
│   with aws_lambda_layer_version.genai,
│   on lambda_layer.tf line 57, in resource "aws_lambda_layer_version" "genai":
│   57: resource "aws_lambda_layer_version" "genai" {
│ 
╵

**Reply by lpm0073** [reply]
Hi Joe and thanks for your support! aws-openai is a long-term project. It’s about all of the minutia that's involved in getting generative AI apps to production; not unlike what you’re running into presently. Terraform is highly imperfect.  At a glance, I think my next step in your case would be
 1. to destroy everything, and then rebuild. It should only take a couple of minutes, and at any rate nothing is persisted so it’s not actually a destructive operation in the classical sense.
 2. If that doesn’t work then you might consider looking for the line in question and then backing up to a commit that precedes your most recent change. 
 3. And if that doesn’t work then I’d manually delete the entire API Gateway api and let Terraform recreate it. 

Hope that helps.

**Reply by jlombard** [reply]
Thanks,  that didn't work for me, but I may be in over my head with the Dev Ops side of this.  I might tack towards using Bedrock for this project since AWS says it's a simple dev path.

**Comment by PirateUnderAPalm** [comment]
Is there a full guide to go from base linux installation through to having this fully running? I worked through the "Getting Started with AWS and Terraform". I saw elsewhere but I can't find it right now about modifying the .env file in the root directory but that file doesn't seem to exist. It seems to want to be created in the Makefile, but introduces a new requirement a Pinecone API key which isn't mentioned in the Readme. When I try to run make init it fails with "recipe commences before first target. Stop." Any guidance would be most appreciated.

**Reply by lpm0073** [reply]
a few suggestions:
1. you might want to ´git pull´ depending on when you cloned/forked the repo, just to ensure that you're running the latest
2. confirming that the command you attempted was `make api-init`. Unfortunately i'm in the situation of "well, it works on my machine". GitHub CoPilot strongly suggests that the error "recipe commences before first target. Stop." is caused by commands being delimited with spaces instead of tabs (it should be tabs). However, in the latest on main i do not find any space-delimited commands. So I don't have any other ideas for the moment.
3. i'd recommend running the commands individually on your command line. you'll find them on row 26 of Makefile. I'm interested in knowing how this works for you: good, bad or otherwise.

You can create the .env manually. Note the format:

```OPENAI_API_ORGANIZATION=org-A-LONG-HASH
OPENAI_API_KEY=sk-AN-EVEN-LONGER-HASH
PINECONE_API_KEY=LOOKS-LIKE-A-GUID
PINECONE_ENVIRONMENT=gcp-SOME-NAME
DEBUG_MODE=True
```

Pinecone is benign for the moment. I'll do a YouTube video in mid December (or January) that makes use of it. until then it's only included in the build as a precursor to ensure that the requirements don't break the Lambda layer or otherwise wreak havoc with anything.

One final thing: GitHub CoPilot suggests that there's a syntax error in the logic to source/create the .env file. it suggests the following modification, which I pushed to 'next' just now and will add it to my next PR which presumably will merge sometime in the next 24 hours.:

```
ifneq ("$(wildcard .env)","")
    include .env
else
    echo -e "OPENAI_API_ORGANIZATION=PLEASE-ADD-ME\nOPENAI_API_KEY=PLEASE-ADD-ME\nPINECONE_API_KEY=PLEASE-ADD-ME\nPINECONE_ENVIRONMENT=gcp-starter\nDEBUG_MODE=True\n" >> .env
endif```

**Reply by lpm0073** [reply]
Please see [Merge pull request #126 Update API setup docs for Pinecone](https://github.com/FullStackWithLawrence/aws-openai/commit/f6cbbec1972401b4319c48c9a01dfec598289262), which should resolve:

- api README outdated setup instructions
- Makefile syntax error in automatically creating .env file if it doesn't already exist

**Reply by PirateUnderAPalm** [reply]
So I pulled the latest version and still ran into issues with the Makefile. It still complained about line 8 which this the one that creates the .env file. I ended up creating it manually and commenting out that line to get it to run. It then started failing because I didn't have npm installed, I installed nvm to then install node and then I was able to run "make api-init" successfully.

**Reply by lpm0073** [reply]
Thanks @PirateUnderAPalm. What do suggest that we do in terms of both the Makefile UX as well as documentation in general in order to a.) make the setup process more clear, and b.) provide better procedural instruction?

Specifically, in hindsight, how might have it been more clear that, for example, node is required for this project? For that matter, Python is required, Terraform is required, AWS CLI, etc. It's a bit challenging for me to attempt to imagine a dev environment that might be missing something that is presently assumed to exist.

**Reply by lpm0073** [reply]
@PirateUnderAPalm I researched the nuances between macOS and Linux, and it seems that wrapping the initialization command in $() might take care of the problem you're mentioning, as follows:

```
ifneq ("$(wildcard .env)","")
    include .env
else
    $(shell echo -e "OPENAI_API_ORGANIZATION=PLEASE-ADD-ME\nOPENAI_API_KEY=PLEASE-ADD-ME\nPINECONE_API_KEY=PLEASE-ADD-ME\nPINECONE_ENVIRONMENT=gcp-starter\nDEBUG_MODE=True\n" >> .env)
endif
```

I [pushed this change to Main](https://github.com/FullStackWithLawrence/aws-openai/commit/a1aa70bae0fd41a4758ccb8e60535bc3061d8ecc) just now.

**Reply by PirateUnderAPalm** [reply]
The existing instructions did make clear that AWS CLI and Terraform are required. Once I got the makefile working it looks like it handled the install of python and a number of other prerequisite packages. I have just tried the updated Makefile and it works as updated on linux (currently running Ubuntu 22.04). I am still fairly new in working with Linux and python (I have more of a Windows C# background) so perhaps some of the things would have been clear to one who has worked in this environment longer.

I have definitely run into the "it works on my machine" dillema myself so I completely understand where you're at and appreciate your help in tracking down the cause.  Overall the documentation is great and I definitely would not have gotten this far without it. I think a section with just a general outline of steps like below linking to the existing sections or new sections as appropriate would help make things clear for other users.

I think an overall outline somewhere similar to the following would have made this easier to follow:
1) Set up AWS account
2) AWS Command Line interface
3) Getting Started with AWS and Terraform
4) Set up OpenAI API Key (Move to 1 or 3 since needed in current 3)
5) Section on Makefile
         a) Not clear npm required unless you open Makefile 
                    i) install nvm
                    ii) install npm
         b) Makefile commands to start API (can you verify the below correct assuming Terraform apply was run in step3?)
	            i) make api-init
	            ii) make api-activate
	            iii) make api-test (check for errors)
	            iv) make api-lint
          c) Makefile Commands to start React App (is the below the intended progression?)
	             i) make client-init
	             ii) make client-lint
	             iii) make client-update
	             iv) make client-run
	             v) make client-build
	             vi) make client-release
6) Access & work with it

Possible Expansions (I plan on trying all of these after the initial project is running so I can probably help with documenting them):

7) Section on how to modify/add code to access one's own GPTs/Assistants through the react App 
8) How to integrate react app (or a scalled down version for one GPT/Assistant) into other websites
9) How to integrate into non-route53 domains

**Reply by lpm0073** [reply]
This is great feedback. thanks! i'll work on refactoring the documentation.

**Comment by PirateUnderAPalm** [comment]
I tried to run terraform apply and I ran into an issue where docker was also needed, I got it resolved but something else to perhaps mention in a refactored documentation.
I am running into an issue where it fails when setting up cloudwatch for the api gateway it says "
cloudwatch logs role must be set in account settings to enable logging" I have set up a role in AWS to enable it for this purpose but I'm unsure where to inject it into the terraform config so that it is integrated when "terraform apply" is run.

**Reply by lpm0073** [reply]
Thanks. I merged [this PR](https://github.com/FullStackWithLawrence/aws-openai/commit/66255579e687933516ebe74d39fd29c3b75cd5b2) and [this one](https://github.com/FullStackWithLawrence/aws-openai/commit/a7e359ccf7c7e65f570bfc201e54bffedaac1ba0) just now.

Feel free to open a PR with any further omissions that you might run across.
